src/main/scala/datalog/execution/ir/IRTreeGenerator.scala

```diff
@@ -98,43 +97,124 @@ class IRTreeGenerator(using val ctx: InterpreterContext)(using JITOptions) {
                     ScanOp(r, DB.Derived, KNOWLEDGE.Known)
                 }): _*
               )
-            }):_*
+            }): _*
           )
       case _ =>
         debug("AST node passed to semiNaiveEval:", () => ctx.storageManager.printer.printAST(ast))
         throw new Exception("Wrong ASTNode received when generating naive IR")
     }
   }
 
-  def generateNaive(ast: ASTNode): IROp[Any] = {
+  /**
+   * A step in the strata generation. At each step, some new relations should
+   * be derived, and some relations should be kept from the previous strata.
+   *
+   * @param toDerive The relations to derive in this step.
+   * @param toKeep   The relations to keep from the previous strata.
+   * @param rules    The rules to derive in this step.
+   */
+  private case class Step(toDerive: Set[RelationId],
+                          toKeep: Set[RelationId],
+                          rules: mutable.Map[Int, ASTNode])
+
+  /**
+   * Stratifies the program into a sequence of steps, where strata are rules
+   * that should be derived simultaneously.
+   *
+   * @param ast The root of the program.
+   * @return A sequence of steps for which relations should be derived.
+   */
+  private def stratify(ast: ASTNode): Seq[Step] = {
     ast match {
       case ProgramNode(ruleMap) =>
-        DoWhileOp(
-          DB.Derived,
-          SequenceOp(OpCode.LOOP_BODY,
-            SwapAndClearOp(),
-            naiveEval(ruleMap)
-          )
-        )
+        val components = ctx.precedenceGraph.scc(ctx.toSolve)
+        val kept = components.scanLeft(Set[Int]())(_ ++ _)
+        components.zip(kept).map((ids, keep) =>
+          val rules = ruleMap.filter((r, _) => ids.contains(r))
+          Step(ids, keep, rules))
       case _ => throw new Exception("Non-root passed to IR Program")
     }
   }
 
+  /**
+   * Generates the IR for the naive evaluation strategy.
+   *
+   * @param ast The root of the program.
+   * @return The IR for the naive evaluation strategy.
+   */
+  def generateNaive(ast: ASTNode): IROp[Any] = {
+    val steps = stratify(ast).map(step =>
+      val keep =
+        SequenceOp(OpCode.OTHER,
+          step.toKeep.toSeq.map(r =>
```

Comment:
Is it necessary to copy all of `keep` into `Derived.New`? I think (but I'm not 100% sure) it should be enough to just copy the final result of each previous strata, which will be stored in `Derived.New`, into `Derived.Known` so that the next strata can read from it. Generally the pattern is that each main program loop will read exclusively from `Known` and write exclusively to `New`, so you might be able to just get away with an extra `SwapAndClearOp` at the end of each strata, since that operation essentially copies all facts from `New` into `Known`. Does that make sense?